### DISCUSSION TITLE
How can I establish communication between OpenDDS and OpenSplice

### DISCUSSION THREAD

**Discussion by estalyncuray** [OpenDDS/OpenDDS#4631] [Q&A] [answered]
I would like to establish a communication between OpenDDS and OpenSplice, I have developed two different projects (dds-opendds and dds-opensplice available on [github](https://github.com/estalyncuray/dds)) 
both contain Publisher (Pub) and Subscriber (Sub) applications.
When I run the ‘Pub’ and ‘Sub’ application from the same project, the data is shared correctly, 
However, when I try to run the ‘Pub’ application from OpenDDS and ‘Sub’ from OpenSplice, the subscriber does not receive the data from the publisher.

Details: 
- OpenDDS.version : 3.28.1, configured in RTPS.
- OpenSplice.version : 6.9.210323OSS, configured in single process DDSI2

I use very "standard" configuration on both sides (to avoid complications).
See the following files for each vendor.

**OpenDDS configuration (rtsp.ini)**
```
[common]
DCPSDefaultAddress=172.18.23.117
DCPSDefaultDiscovery=DEFAULT_RTPS
DCPSGlobalTransportConfig=myconfig
DCPSDebugLevel=10
DCPSTransportDebugLevel=1

[domain/0]
DiscoveryConfig=uni_rtps

[config/myconfig]
transports=the_rtps_transport

[rtps_discovery/uni_rtps]
SedpMulticast=1
ResendPeriod=2
SpdpRtpsRelayAddress=172.18.23.117:
SedpRtpsRelayAddress=172.18.23.117:

[transport/the_rtps_transport]
multicast_interface=Ethernet
transport_type=rtps_udp
use_multicast=1
DataRtpsRelayAddress=172.18.23.117:
``` 

**OpenSplice configuration (ospl.xm)**

```
<OpenSplice>
  <Domain>
    <Name>ospl_sp_ddsi</Name>
    <Id>0</Id>
    <SingleProcess>true</SingleProcess>
    <Service name="ddsi2">
      <Command>ddsi2</Command>
    </Service>
    <Report append="false" verbosity="info" />
  </Domain>
  <DDSI2Service name="ddsi2">
    <General>
      <NetworkInterfaceAddress>AUTO</NetworkInterfaceAddress>
      <AllowMulticast>true</AllowMulticast>
      <EnableMulticastLoopback>true</EnableMulticastLoopback>
      <CoexistWithNativeNetworking>false</CoexistWithNativeNetworking>
    </General>
    <Discovery>
      <ParticipantIndex>auto</ParticipantIndex>
      <SPDPInterval>30 s</SPDPInterval>
    </Discovery>
    <Compatibility>
      <StandardsConformance>lax</StandardsConformance>
    </Compatibility>
	<Tracing>
        <AppendToFile>false</AppendToFile>
        <Verbosity>fine</Verbosity>
    </Tracing>    
  </DDSI2Service>
</OpenSplice>
``` 

**Then, the important note is:**

In the ‘RTI Connext DDS View’ I can see that the communication related to the OpenDDS publisher and OpenSplice subscriber is established correctly, this means that I am missing some configuration details.

What did I miss?
What am I doing "wrong"?
If anyone can help me, I would be fully grateful!

**Comment by iguessthislldo** [comment]
If OpenSplice isn't receiving, it sounds like it might be a [data representation QoS](https://opendds.readthedocs.io/en/master/devguide/xtypes.html#xtypes-data-representation) mismatch. If that's the case you'll have to either switch the OpenSlice reader to also accept XCDR2 if they support that or set the OpenDDS writer to use XCDR1. There's a C++ example of setting a OpenDDS writer to XCDR1 in the section I linked.

**Reply by estalyncuray** [reply]
Thanks for the answer, 
I have created a Qos data representation in OpenDDS, before adding the Qos to the Java project I change the IDL structure as detailed below,
so the problem is limited to accept XCDR1 (I am not sure about it).

**A.idl file**
```
module com{
	module example {
		module dds {
			module a {
			@topic
			@::OpenDDS::data_representation(XCDR1)
				struct A {
					@key string id;
					string name;
				};
			};
		};
	};
};
``` 
**A.mpc file**
```
project: dcps_java, dcps_rtps, dcps_rtps_udp {
  idlflags += -Wb,stub_export_include=A_Export.h \
    -Wb,stub_export_macro=A_Export \
	-Cw
  dcps_ts_flags += -Wb,export_macro=A_Export \
  -Cw
  idl2jniflags += -Wb,stub_export_include=A_Export.h \
    -Wb,stub_export_macro=A_Export \
	-Cw
  dynamicflags += A_BUILD_DLL

  specific {
    jarname = DDS_A_types
  }

 TypeSupport_Files {
    A.idl
  }
}
``` 

After modified the IDL, I added the respective QoS to the topicQos and dataWriterQos representation value.

**Instantiate DataRepresentationQoS**
```
dataRepresentationIdSeqHolder.value = new short[1];
dataRepresentationIdSeqHolder.value[0] = XCDR_DATA_REPRESENTATION.value;

``` 
**Instantiate TopicQos**
```
topicQosHolder.value.representation.value = dataRepresentationIdSeqHolder.value;
``` 
**Instantiate DataWriterQos**
```
dataWriterQosHolder.value.representation.value = dataRepresentationIdSeqHolder.value;
``` 

It still not working, What am I doing "wrong"?

**Reply by jrw972** [reply]
Please take a packet capture and post it.

**Reply by estalyncuray** [reply]
Sorry for the late replay.

Below I attach a packet of information obtained from wireshark. 
Under the wireshark, there is a lot of information that I am not able to understand (the DDS world is new to me). 
I can see packets that are using the RTPS protocol, each packet contains individual information related to OpenDDS and OpenSplice.

**OpenDDS**

![OpenDDS-wireshark](https://github.com/OpenDDS/OpenDDS/assets/123993751/3e93519e-9a3d-4ef7-8176-d66c20281901)

**OpenSplice**
![OpenSplice-wireshark](https://github.com/OpenDDS/OpenDDS/assets/123993751/5597320a-7aa0-4ace-869c-fbc4d5fd05f4)

**Reply by jrw972** [reply]
Post the actual captures and not pictures.

**Reply by estalyncuray** [reply]
Wireshark packets filter by rtps.

**Reply by jrw972** [reply]
The packet capture has the following participants:
- 0103745d22207ebb86c4c4fa (OpenDDS) 192.168.1.5 239.255.0.1:7400
- 3dfd07790000000400000001 (OpenSplice) 192.168.1.5 239.255.0.1:7400
- 3dfd07790000005800000001 (OpenSplice) 192.168.1.5 239.255.0.1:7400
- 3dfd07790000007d00000001 (OpenSplice) 192.168.1.5 239.255.0.1:7400
- 3dfd0779000001b900000001 (OpenSplice) 192.168.1.5 239.255.0.1:7400
- 57ccf67c0000005b00000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000000b00000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000003000000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000005e00000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000011600000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000012500000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000013300000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000014100000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000015d00000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650
- 57ccf67c0000016b00000001 (OpenSplice) 172.18.23.65 239.255.0.1:7650

I'll ignore participants with GuidPrefix 57* because they are appear to be on a different domain and network.

The first SPDP announcement from the OpenDDS participants has sequence number 3.

Please take the capture again and start the capture before the OpenDDS participant has started.

**Reply by estalyncuray** [reply]
I upload the packet capture.

**Reply by jrw972** [reply]
Thanks.

What I expect to see is OpenDDS sending an SPDP message to a discovered participant when it first sees one from the remote.
This is not happening.
The OpenSplice participants seem to not be getting SPDP messages either.
Otherwise, SEDP would start.

I would try two things.
First, run an example that just uses OpenDDS and rtps like `OpenDDS6/DevGuideExamples/DCPS/Messenger/run_test.pl --rtps` to see if that works.
Second, check your network settings to make sure that a firewall is not causing a problem.

**Reply by mitza-oci** [reply]
Could it be the same issue as https://github.com/OpenDDS/OpenDDS/issues/665 ?

**Reply by estalyncuray** [reply]
@jrw972 I have tried what you suggest and it works fine (attached result - result_run-test.txt, packet_run-test.pcapng). I have also tried to run the Publisher 'Pub' and Subscriber 'Sub' of the Java project (dds-opendds), taking into account the rtps protocol and it still works correctly (attached packet capture- packet_dds-opendds.pcapng).

The problem comes when I try to run the Java projects 'Pub' of OpenDDS (dds-opendds) and 'Sub' of OpenSplice (dds-opensplice).

**Reply by estalyncuray** [reply]
@mitza-oci I am not sure if it is related to the same problem, but I will try to change OpenDDS-3.28.1 (dds\DCPS\RTPS\MessageTypes.h) protocol back to version 2.1.

**Reply by estalyncuray** [reply]
@mitza-oci I have modified the protocol and compiled the RTPS project, the library 'OpenDDS_Rtps.dll' has been regenerated and added to my project, but it still not working. Is it necessary to compile the complete DDS_TAOv2.sln project again or is the RTPS.sln project enough?

**Reply by jrw972** [reply]
I think you need to rebuild both the rtps dll and the rpts udp transport dll.

The OpenDDS-OpenDDS captures look strange.  The captures only show multicast traffic.  There is most likely some unicast traffic that is missing and that would help explain what is going on.

**Reply by estalyncuray** [reply]
I am using the configuration detailed below, to try to communicate OpenDDS and OpenSplice.

OpenDDS configuration
```
[common]
DCPSGlobalTransportConfig=$file
DCPSDebugLevel=10
DCPSTransportDebugLevel=5
DCPSTypeObjectEncoding=ReadOldFormat

[domain/example]
DomainId=0
DiscoveryConfig=uni_rtps

[rtps_discovery/uni_rtps]
SedpMulticast=1
ResendPeriod=2

[transport/the_rtps_transport]
multicast_interface=Ethernet
transport_type=rtps_udp
use_multicast=1
max_message_size=655510
rcv_buffer_size=6553600
``` 

OpenSplice does not mention anything about how to configure rtps protocol in the ospl.xml file, here is the [link ](https://download.zettascale.online/www/docs/Vortex/html/ospl/DeploymentGuide/guide.html), so I set the configuration as follows. 
```
<OpenSplice>
  <Domain>
    <Name>example</Name>
    <Id>0</Id>
    <SingleProcess>true</SingleProcess>
    <Service name="ddsi2">
      <Command>ddsi2</Command>
    </Service>
    <Report append="false" verbosity="INFO" />
  </Domain>
  <DDSI2Service name="ddsi2">
    <General>
      <NetworkInterfaceAddress>AUTO</NetworkInterfaceAddress>
      <AllowMulticast>true</AllowMulticast>
      <EnableMulticastLoopback>true</EnableMulticastLoopback>
      <CoexistWithNativeNetworking>false</CoexistWithNativeNetworking>
    </General>
    <Discovery>
      <ParticipantIndex>auto</ParticipantIndex>
      <SPDPInterval>30 s</SPDPInterval>
    </Discovery>
    <Compatibility>
      <StandardsConformance>lax</StandardsConformance>
    </Compatibility>
  </DDSI2Service>
</OpenSplice>
``` 

It is the configuration correct ?

**Reply by jrw972** [reply]
I would omit `DCPSTypeObjectEncoding` as I don't think you are interacting with older OpenDDS version.

**Comment by jrw972** [comment]
Why are you specifying `DCPSDefaultAddress`, `SpdpRtpsRelayAddress`, `SedpRtpsRelayAddress`, and `DataRtpsRelayAddress`?  If you are actually using the RtpsRelay, then the relay addresses need ports and you also need to generally enable the RtpsRelay.

**Reply by estalyncuray** [reply]
Sorry for the late reply, 

I am a bit new working with DDS services like OpenDDS, so maybe I made a mistake when setting the run-time configuration details.

- Is there a standard configuration to communicate OpenDDS and OpenSplice ?
- It is necessary to enable multicast?

**Reply by jrw972** [reply]
I'm not aware of a standard configuration.

It shouldn't be necessary unless OpenSplice only supports multicast.

**Comment by estalyncuray** [comment, marked as answer]
Finally, after trying many suggestions from all of you (Thank you), I have managed to solve the problem related to the communication between OpenDDS and OpenSplice. I am going to list some steps I did to be able to communicate both DDS.

**Communication between Publisher (OpenDDS) and Subscriber (OpenSplice)**

**Publisher (OpenDDS)**
1 - (Optional) It is not necessary to modify the rpts protocol version from 2.4 to 2.1 in the dds\DCPS\RTPS\MessageTypes.
2 - (Mandatory) Add the ‘representation.value’ to ‘_XCDR_DATA_REPRESENTATION.value_’ in the _dataWriterQoS_ of the publisher. 
```java
dataWriterQosHolder.value = DATAWRITER_QOS_DEFAULT.get();
publisher.get_default_datawriter_qos(dataWriterQosHolder);
dataWriterQosHolder.value.representation.value = new short[1];
dataWriterQosHolder.value.representation.value[0] = XCDR_DATA_REPRESENTATION.value;
publisher.set_default_datawriter_qos(dataWriterQosHolder.value);
```  
3 - (Optional) Modify _Topic.idl_ structure by adding the attribute ‘_@::OpenDDS::data_representation(XCDR1)_’.
```
module com {
	module example {
		@topic
		@::OpenDDS::data_representation(XCDR1)
		struct MyTopic {
			@key string id;
			string name;
		};
	};
};
``` 
3.1 - (Mandatory if step 3 is followed) recompile Topic.idl using the Topic.mpc file below.
```
project: dcps_java {
  idlflags += -Wb,stub_export_include=MyTopic_Export.h \
    -Wb,stub_export_macro=MyTopic_Export \
	-Cw 
  dcps_ts_flags += -Wb,export_macro=MyTopic_Export \
  -Cw
  idl2jniflags += -Wb,stub_export_include=MyTopic_Export.h \
    -Wb,stub_export_macro=MyTopic_Export \
	-Cw 
  dynamicflags += MYTOPIC_BUILD_DLL

 TypeSupport_Files {
    MyTopic.idl
  }
}
``` 
3.2 - (Mandatory if step 3 is followed) Add ‘representation.value’ to ‘XCDR_DATA_REPRESENTATION.value’ in the editor _topicQoS_. 
```java
topicQosHolder.value = TOPIC_QOS_DEFAULT.get();
domainParticipant.get_default_topic_qos(topicQosHolder);
topicQosHolder.value.representation.value = new short[1];
topicQosHolder.value.representation.value[0] = XCDR_DATA_REPRESENTATION.value;
domainParticipant.set_default_topic_qos(topicQosHolder.value);
``` 
4 - (Mandatory) Add ‘DCPSDefaultAddress=X.X.X.X.X’ to the OpenDDS configuration file (rtps.ini).
```
[common]
DCPSDefaultAddress=172.18.23.76
DCPSGlobalTransportConfig=$file

[domain/0]
DiscoveryConfig=uni_rtps

[rtps_discovery/uni_rtps]
SedpMulticast=1
ResendPeriod=2

[transport/the_rtps_transport]
transport_type=rtps_udp
use_multicast=1

``` 
**Subscriber (OpenSplice)**
1 - (Mandatory) Add the '_Discovery_' tag to the ospl.xml file, otherwise the communication will not work. Even if the previous steps are correctly done.
```xslt
<OpenSplice>
    <Domain>
        <Name>MyTopicPartition</Name>
        <Id>0</Id>
        <SingleProcess>true</SingleProcess>
        <Service name="ddsi2">
            <Command>ddsi2</Command>
            <FailureAction>restart</FailureAction>
        </Service>
        <Lease>
            <ExpiryTime update_factor="0.2">50.0</ExpiryTime>
        </Lease>
        <Daemon>
            <Heartbeat>
                <ExpiryTime update_factor="0.25">20.0</ExpiryTime>
            </Heartbeat>
        </Daemon>
        <Report append="false" verbosity="INFO"/>
    </Domain>
    <DDSI2Service name="ddsi2">
        <General>
            <NetworkInterfaceAddress>AUTO</NetworkInterfaceAddress>
            <AllowMulticast>true</AllowMulticast>
            <EnableMulticastLoopback>true</EnableMulticastLoopback>
            <CoexistWithNativeNetworking>false</CoexistWithNativeNetworking>
        </General>
        <Discovery>
            <ParticipantIndex>none</ParticipantIndex>
            <SPDPInterval>30 s</SPDPInterval>
        </Discovery>
        <Internal>
            <SquashParticipants>true</SquashParticipants>
        </Internal>
        <Compatibility>
            <StandardsConformance>lax</StandardsConformance>
        </Compatibility>
		<Tracing>
            <AppendToFile>false</AppendToFile>
            <Verbosity>finest</Verbosity>
        </Tracing>
    </DDSI2Service>
</OpenSplice>
``` 
**Communication between Publisher (OpenSplice) and Subscriber (OpenDDS)**

1 - (Optional) It is not necessary to add ‘_XCDR_DATA_REPRESENTATION.value_’ in the subscriber's topicQoS or in the dataReaderQos of the OpenDDS.
